File: web-local/src/routes/dev/numbers/smallest-precision-magnitude.ts
// returns the smallest order of magnitude to which a number
// has precision -- basically, the smallest OoM that has a non-zero
// digit. Returm NaN if number is NaN or infinite
export const smallestPrecisionMagnitude = (x: number): number => {
  if (!isFinite(x)) return NaN;
  if (x === 0) return 0;
  // if the number is not an integer, find the smallest fractional digit
  if (!Number.isInteger(x)) {
    // x = x - Math.round(x);
    let e = 1;
    let p = 0;
    while (Math.round(x * e) / e !== x) {
      e *= 10;
      p++;
    }
    return -p;
  }

  let e = 10;
  let p = 0;
  while (Math.round(x / e) * e === x) {
    e *= 10;
    p++;
  }
  return p;
};

const testCases: [number, number][] = [
  [Infinity, NaN],
  [-Infinity, NaN],
  [NaN, NaN],
  [0, 0],
  [-0, 0],
  [0.1, -1],
  [0.000000001, -9],
  [0.000027391, -9],
  [0.973427391, -9],
  [123, 0],
  [1230, 1],
  [123000000000, 9],
  [1239347293742974, 0],
  [6000000000, 9],
  [-0.1, -1],
  [-0.000000001, -9],
  [-0.000027391, -9],
  [-0.973427391, -9],
  [-123, 0],
  [-1230, 1],
  [-123000000000, 9],
  [-1239347293742974, 0],
  [-6000000000, 9],
  [710.7237956, -7],
  [-710.7237956, -7],

  [79879879710.7237, -4], // NOTE: most digits representable in js
  [-79879879710.7237, -4], // NOTE: most digits representable in js
];

export const runTestsmallestPrecisionMagnitude = () => {
  const caseResults = testCases.map((tc) => {
    const out = smallestPrecisionMagnitude(tc[0]);
    const correct = Number.isFinite(tc[0]) ? out === tc[1] : Number.isNaN(out);
    return [correct ? "pass" : "fail", tc, out];
  });

  const overallPass = caseResults.reduce(
    (prev, tc) => prev && tc[0] === "pass",
    true
  );

  if (!overallPass) {
    console.error("smallestPrecisionMagnitude test FAIL:", caseResults);
  } else {
    console.log("smallestPrecisionMagnitude test pass:", caseResults);
  }
};

File: runtime/pkg/formatter/smallest_precision_magnitude.go
package formatter

import (
	"fmt"
	"math"
	"strconv"
	"strings"

	"golang.org/x/exp/constraints"
)

// returns the smallest order of magnitude to which a number has precision --
// basically, the smallest OoM that has a non-zero digit.
func smallestPrecisionMagnitude[T Number](x T) int {
	if isUnsigned(x) {
		return smallestPrecisionMagnitudeInt(uint64(x))
	}
	if isInteger(x) {
		return smallestPrecisionMagnitudeInt(int64(x))
	}
	return smallestPrecisionMagnitudeFloat(float64(x))
}

func smallestPrecisionMagnitudeInt[T int64 | uint64](x T) int {
	if x == 0 {
		return 0
	}

	var e T = 10
	p := 0
	for x/e*e == x {
		e *= 10
		p++
	}
	return p
}

func smallestPrecisionMagnitudeFloat(x float64) int {
	if math.IsNaN(x) || math.IsInf(x, 0) {
		return 0 // This should never happen
	}

	if x == 0 {
		return 0
	}

	if math.Abs(x) > 1e280 {
		return smallestPrecisionMagnitudeLargeNumber(x)
	}

	e := 1.0
	p := 0
	for math.Round(x*e)/e != x && p < 324 {
		e *= 10
		p++
	}
	return -p
}

func smallestPrecisionMagnitudeLargeNumber(n float64) int {
	s := fmt.Sprintf("%e", n)
	eIndex := strings.Index(s, "e")
	dotIndex := strings.Index(s, ".")
	exp, _ := strconv.Atoi(s[eIndex+1:])
	digitsAfterDot := eIndex - dotIndex - 1
	i := exp - digitsAfterDot
	return i
}

type Number interface {
	constraints.Integer | constraints.Float
}

func asNumber[T Number](x any) (T, bool) {
	switch x := x.(type) {
	case int:
		return T(x), true
	case int8:
		return T(x), true
	case int16:
		return T(x), true
	case int32:
		return T(x), true
	case int64:
		return T(x), true
	case uint:
		return T(x), true
	case uint8:
		return T(x), true
	case uint16:
		return T(x), true
	case uint32:
		return T(x), true
	case uint64:
		return T(x), true
	case float32:
		return T(x), true
	case float64:
		return T(x), true
	}
	return 0, false
}

func isUnsigned(x any) bool {
	switch x.(type) {
	case uint, uint8, uint16, uint32, uint64:
		return true
	}
	return false
}

func asUnsigned(x any) (uint64, bool) {
	switch x := x.(type) {
	case uint:
		return uint64(x), true
	case uint8:
		return uint64(x), true
	case uint16:
		return uint64(x), true
	case uint32:
		return uint64(x), true
	case uint64:
		return x, true
	}
	return 0, false
}

func isInteger(x any) bool {
	switch x.(type) {
	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
		return true
	}
	return false
}

func asInteger(x any) (int64, bool) {
	switch x := x.(type) {
	case int:
		return int64(x), true
	case int8:
		return int64(x), true
	case int16:
		return int64(x), true
	case int32:
		return int64(x), true
	case int64:
		return x, true
	}
	return 0, false
}

func isFloat(x any) bool {
	switch x.(type) {
	case float32, float64:
		return true
	}
	return false
}

func asFloat(x any) (float64, bool) {
	switch x := x.(type) {
	case float32:
		return float64(x), true
	case float64:
		return x, true
	}
	return 0, false
}
